### PR TITLE
wip: shader atlas

### PR DESCRIPTION
**Pull request by luan**
Note, for shader auto reloading:

```
BEVY_ASSET_ROOT=$PWD cargo run -p ryot_compass --features bevy/dynamic_linking --features bevy/file_watcher --features bevy/embedded_watcher --features lmdb
```

### CODE REVIEW

File: ryot/src/appearances/appearances.proto [hunk 1]
```diff
@@ -99,7 +99,7 @@ message SpriteInfo {
 	optional uint32 pattern_height = 2;
 	optional uint32 pattern_depth = 3;
 	optional uint32 layers = 4;
-	repeated uint32 sprite_id = 5;
```
Comment (lgrossi) [inline ryot/src/appearances/appearances.proto:102]: Noice, I think I'll add a linear ticket to review this proto and put better and more intuitive names on it.

File: ryot/src/bevy_ryot/drawing/systems/mod.rs [hunk 2]
```diff
@@ -80,29 +80,23 @@ pub fn get_top_most_visible_for_tile(
...
-        Or<(
-            Changed<Visibility>,
-            Added<Visibility>,
-            Changed<TilePosition>,
-        )>,
-    ),
->;
+type ElevationFilter = (
+    With<TileComponent>,
+    Or<(Changed<Visibility>, Changed<TilePosition>)>,
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/drawing/systems/mod.rs:85]: Changed includes added?

File: ryot/src/bevy_ryot/sprites.rs [hunk 3]
```diff
@@ -227,262 +186,221 @@ pub(crate) fn prepare_sprites<C: PreloadedContentAssets>(
...
+pub struct LoadAppearanceEvent(pub AppearanceDescriptor);
+
+pub(crate) fn load_from_entities_system(
+    query: Query<&AppearanceDescriptor, Changed<AppearanceDescriptor>>,
+    loaded_appearances: Res<LoadedAppearances>,
+    mut events: EventWriter<LoadAppearanceEvent>,
+) {
+    query
+        .iter()
+        .unique()
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/sprites.rs:229]: Does bevy give you the same entity more than once in a query?

File: ryot/src/bevy_ryot/sprites.rs [hunk 4]
```diff
@@ -227,262 +186,221 @@ pub(crate) fn prepare_sprites<C: PreloadedContentAssets>(
...
-            let (sprite, anim) = match load_desired_appereance_sprite(
-                descriptor.group,
-                descriptor.id,
-                descriptor.frame_group_index(),
-                animation_sprite,
-                direction.as_deref(),
+    let descriptors = events
+        .read()
+        .map(|LoadAppearanceEvent(descriptor)| descriptor)
+        .filter(|descriptor| descriptor.id != 0)
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/sprites.rs:248]: nbc: filter_map instead?

File: ryot/src/bevy_ryot/sprites.rs [hunk 5]
```diff
@@ -227,262 +186,221 @@ pub(crate) fn prepare_sprites<C: PreloadedContentAssets>(
...
+    }
+    debug!(
+        "Loading ids: {:?}",
+        descriptors
+            .iter()
+            .map(|descriptor| descriptor.id)
+            .collect::<Vec<u32>>()
+    );
+
+    let descriptor_infos: HashMap<AppearanceDescriptor, SpriteInfo> = descriptors
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/sprites.rs]: nbc: descriptor_sprite_infos?

File: ryot/src/sprites/position.rs [hunk 6]
```diff
@@ -374,100 +378,74 @@ impl DivAssign<f32> for Sector {
...
-    for (_entity, _tile_pos, elevation, _layer, _loaded_sprite, transform, _movement, children) in
-        query.iter_mut()
-    {
-        if let Some(children) = children {
-            for child in children.iter() {
-                if let Ok(mut text) = children_query.get_mut(*child) {
-                    text.text = format!("{:.02} [{}]", 1000. * transform.translation.z, elevation);
-                }
+#[cfg(feature = "bevy")]
+pub fn animate_sprite_position(
```
Comment (lgrossi) [inline ryot/src/sprites/position.rs:403]: nbc: move_sprites_with_animation?

File: ryot/src/bevy_ryot/sprites.rs [hunk 7]
```diff
@@ -227,262 +186,221 @@ pub(crate) fn prepare_sprites<C: PreloadedContentAssets>(
...
+                .try_insert(key.clone(), key.default_state())
+                .ok();
         }
-    }
+    });
 }
 
-pub(crate) fn update_sprite_system<C: ContentAssets>() {}
+#[derive(Event)]
+pub struct LoadAppearanceEvent(pub AppearanceDescriptor);
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/sprites.rs]: NBC: For the sake of my OCD can we call this `LoadAppeareanceCommand`?

File: ryot/src/bevy_ryot/mod.rs [hunk 8]
```diff
@@ -164,10 +164,13 @@ impl<C: PreloadedContentAssets + Default> Plugin for ContentPlugin<C> {
             .add_systems(
                 Update,
                 (
-                    sprites::load_sprite_system::<C>
+                    sprites::process_load_events_system::<C>
+                        .pipe(sprites::load_sprite_system::<C>)
+                        .pipe(sprites::store_loaded_appearances_system)
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/mod.rs:162]: :heart:

File: ryot/src/bevy_ryot/sprites.rs [hunk 9]
```diff
@@ -211,340 +154,235 @@ pub(crate) fn load_sprite_texture<C: ContentAssets>(
...
-                ..default()
+    layouts: Res<Assets<TextureAtlasLayout>>,
+    sprite_meshes: Res<SpriteMeshes>,
+    mut materials: ResMut<Assets<SpriteMaterial>>,
+    asset_server: Res<AssetServer>,
+) -> Option<(
+    HashMap<AppearanceDescriptor, SpriteInfo>,
+    HashMap<u32, LoadedSprite>,
+)> {
+    let descriptor_infos = match descriptor_infos {
```
Comment (lgrossi) [inline ryot/src/bevy_ryot/sprites.rs:292]:
```rust
pub(crate) fn load_sprite_system<C: ContentAssets>(
    In(descriptor_infos): In<Option<HashMap<AppearanceDescriptor, SpriteInfo>>>,
    content_assets: Res<C>,
    layouts: Res<Assets<TextureAtlasLayout>>,
    sprite_meshes: Res<SpriteMeshes>,
    mut materials: ResMut<Assets<SpriteMaterial>>,
    asset_server: Res<AssetServer>,
) -> Option<(
    HashMap<AppearanceDescriptor, SpriteInfo>,
    HashMap<u32, LoadedSprite>,
)> {
    descriptor_infos.map(|descriptor_infos| {
        (
            descriptor_infos.clone(),
            descriptor_infos
                .iter()
                .map(|(descriptor, sprite_info)| (descriptor.group, sprite_info))
                .group_by(|(group, _)| *group)
                .into_iter()
                .map(|(group, group_iter)| {
                    let sprite_ids: Vec<u32> = group_iter
                        .flat_map(|(_, sprite_info)| sprite_info.sprite_ids.clone())
                        .collect();
                    (group, sprite_ids)
                })
                .flat_map(|(group, sprite_ids)| {
                    load_sprites(
                        group,
                        sprite_ids,
                        &content_assets,
                        &layouts,
                        &sprite_meshes,
                        &mut materials,
                        &asset_server,
                    )
                })
                .map(|sprite| (sprite.sprite_id, sprite))
                .collect(),
        )
    })
}
```

File: ryot/src/bevy_ryot/drawing/systems/mod.rs (hunk 2 above)
Comment (luan) [reply ryot/src/bevy_ryot/drawing/systems/mod.rs:85]: Yup!

File: ryot/src/appearances/appearances.proto (hunk 1 above)
Comment (lgrossi) [reply ryot/src/appearances/appearances.proto:102]: https://linear.app/ryot/issue/RYO-60/review-appearancesproto-names

**Pull request merged**